### ISSUE TITLE
Proposal: Get rid of all serialization in Rhino

### ISSUE THREAD

**Issue by gbrail** [mozilla/rhino#704]
Rhino tries to make all JavaScript objects and other objects Serializable, because it's always done that.

I think that we should consider getting rid of Serialization support for a future release for a few reasons:

1. There is a pretty big category of security holes that are created when things are Serializable.
2. Keeping things Serializable is a maintenance burden and makes it harder for us to change internal structures.
3. Others in the Java community are considering removing it from Java some day.
4. Most other JavaScript engines handle this using JSON.

Can people who actually Serialize and de-serialize Rhino objects please comment here on whether  you are using Serialization, and how?

**Comment by rbri** [comment]
HtmlUnit does not use serialization of JavaScript objects.
Everything that gives us more freedom to change the code base is very welcome.

**Comment by gbrail** [comment]
I'd actually still like to do this -- serialization is being more and more generally acceptable as being terrible for interoperability and security. Unless someone has code that will break I am going to create a new PR soon.

**Comment by michbarsinai** [comment]
Combined with continuations, serialization is a super-powerful tool, as it allows a single machine to run virtually unlimited amount of threads. This is by capturing and serializing continuations, and storing them in a DB. Think of a server that runs multiple processes based on external events. I've heard of at least one system that's doing this using Rhino. I know there was also an experimental project that used serialized continuations to move a computation between servers, but I'm not sure if it ever reached production.

In BPjs we depend on serialized continuations for re-trying computations.

I of course understand the complications serializations add, but given the power they give to Rhino, I'd be happy to see them stay.

**Comment by tonygermano** [comment]
The Continuations API announcement specifically mentioned serialization https://web.archive.org/web/20210226081112/https://developer.mozilla.org/en-US/docs/Mozilla/Projects/Rhino/New_in_Rhino_1.7R2

Serializing Continuations has also appeared in a few SO questions that I came across:
https://stackoverflow.com/questions/8616512/interpreting-javascript-in-java-with-rhino-pausing-resuming-scripts
https://stackoverflow.com/questions/5221175/is-rhino-the-only-interpreter-with-support-for-sandboxing-and-serializable-conti

**Comment by p-bakker** [comment]
@gbrail based on the discussion here and in the [Next steps for Rhino after the v1.7.14 release](https://github.com/mozilla/rhino/discussions/972) discussion, it looks like serialization, at least in the context of Continuations is something that is still desirable.

Not being an expert (or even a rookie in this area), but what our options? Do we have to keep serialization indefinetly, or are there other options?

Based on your initial coming in this issue:
> and makes it harder for us to change internal structures.
Can we eliminate this hurdle by saying that we dont support loading serialized stuff with other version as with which they were written?

>  Most other JavaScript engines handle this using JSON.
I don't think other engines support continuations, which I assume require also storing the callstack, but would that not also be something we could offer as JSON? Not sure how would work with POJOs though.

**Comment by jgomer2001** [comment]
> Combined with continuations, serialization is a super-powerful tool

Agree 100%

> I don't think other engines support continuations

Agree as well. 
Except for Kotlin coroutines and Rhino, I think there are no other means to support the concept of resumable computations in the JVM.

Currently we are incubating a project consisting of a framework for web flows. At the heart of it we have Rhino serializable continuations. They allow us to restore the state of a given flow regardless the end-user hits a different node (server) after every form submission (think of multiple java containers with the same stateless webapp deployed).

Having such a bad reputation, I tried to see if I could switch from Java serialization to a different serialization mechanism (like [kryo](https://github.com/EsotericSoftware/kryo)) but I couldn't get any far:  I noticed several core classes provide their custom `readObject` and `writeObject` methods so that introduces a "logical" dependency on how I would have to do serialization with a different tool. 

I wonder if in the future you would like to take an approach for serialization so it isn't coupled to the Java serialization API. Maybe this would release some of the pain that @gbrail regarded at the beginning of this issue (numeral 2)

PS: if you have any suggestion for me (so that I can use my own serialization tool), let me know. Ideally I need to support serialization of arbitrary objects that hold flows' state, ie. I don't know the Classes a priori.

**Comment by michbarsinai** [comment]
@jgomer2001 in BPjs, we subclassed `ScriptableOutputStream`, and replace objects with object stubs before they go into the stream. We do the reverse process on de-serializations. 

See here:

https://github.com/bThink-BGU/BPjs/blob/2650ad1d588d1aaf83740b1b1be2d0d3dc84bda5/src/main/java/il/ac/bgu/cs/bp/bpjs/bprogramio/BPJSStubOutputStream.java#L48

Credit for the original idea goes to @szegedi, which used it in Spring JSFlow (https://github.com/szegedi/spring-web-jsflow/blob/master/src/main/java/org/szegedi/spring/web/jsflow/support/FlowStateSerializer.java)

**Comment by jgomer2001** [comment]
Pretty interesting, thank you!

**Comment by bunkenburg** [comment]
I agree. Serialisation goes together with continuations. I use both together in my toy project https://www.yellow.cat/. 

A different serialisation mechanism would not work for me, because I have host objects in the JavaScript that are Java objects, so they have to be serialised too.

**Comment by aardvark179** [comment]
So, to be able to serialise Java script object graphs we do need to be able to handle most of the things that other serialisation frameworks try to avoid, specifically we cannot guarantee that the graph is acyclic.

If we provided better async support then tThat might cover some use cases, it would certainly allow a process to block on a promise and then have that resolved later to resume compilation, but that doesn't cover serialisation and passing to another JVM.

I'm curious how well that scenario works in general anyway. I don't think we have nearly enough testing to properly cover generators and other modern features, what happens with a non empty micro task queue.

**Comment by tonygermano** [comment]
It's never been clear to me how async should work. I use Rhino embedded where there can be dozens or even hundreds of short executing scripts running concurrently. What should happen if the script reaches the end without explicitly blocking on all outstanding requests (whether a Promise or `setTimeout`)?

**Comment by aardvark179** [comment]
So promises are enquired as micro tasks, and you can drain that queue when the context is exited. If you have promises that cannot be fulfilled then you could hold on to that context until they can be.

I have a prototype async implementation, and maybe I should add something under examples to show how that could be used.

**Comment by gbrail** [comment]
If everything is contained within Rhino and the scripts that it's running
then the existing Promise implementation and its microtask queue should
Just Work because the script won't exit until all the pending microtasks
are resolved.

Once you have Promises that are pending until something happens outside the
script, such as timers or network I/O, then we need to wrap Rhino in some
larger event loop that will call back into Rhino when the thing happens.
This is actually not all that complicated.

The "shell" module has an implementation of setTimeout that works this way
already, and back when I wrote Trireme 10+ years ago that's what we did
there to support the Node.js API.

**Comment by tonygermano** [comment]
The product I'm referring to is https://github.com/nextgenhealthcare/connect and its forks. https://github.com/OpenIntegrationEngine/engine is the one that I am currently working on. The server application provides a platform and the actual scripts are authored by users.

The user has full access to java classes, and most things that would benefit from async are all outside of rhino, HTTP requests, JDBC connections, File I/O, etc...
